### DISCUSSION TITLE
how tinyusb developers had thought about using tud_audio_fb_set()

### DISCUSSION THREAD

**Discussion by mastupristi** [hathach/tinyusb#2499] [Q&A]
In my project I need to calculate the feedback value in a regular way. For this reason I have a software timer from FreeRTOS, which performs the calculation every 10ms.  The feedback thus computed I use in the `tud_audio_fb_set()` call, which eventually if the endpiont is free calls the `xfer`.
However, `tud_task()` also internally calls the `xfer` on the same EP. This can cause a race condition, and trigger the assert https://github.com/hathach/tinyusb/blob/master/src/device/usbd.c#L1236

In such cases how should one proceed?
How tinyusb developers had thought about using `tud_audio_fb_set()`?
Is it right that `tud_audio_fb_set()` also calls the transfer in addition to setting the feedback value in the internal structures?

regards
Max

**Comment by HiFiPhile** [comment]
Currently `tud_audio_fb_set()` doesn't has racing protection like CDC class does, I'll add it later.

**Reply by HiFiPhile** [reply]
3e393d0 added mutex protection for `audiod_fb_send()`.
